#444 - overtime application process

kamilpiech97 opened the pull request:
Connected with #444.

Added the ability to add an overtime application, where the employee can add the date and time `from` and `to`, `type` of preferred overtime settlement. Added tests.

Preview:

[Screencast from 10-06-24 13:47:25.webm](https://github.com/blumilksoftware/toby/assets/29115024/cc631290-154b-4242-9671-98709ed5322d)

krzysztofrewak commented on app/Actions/OvertimeRequest/AcceptAsTechnicalAction.php:
```diff
@@ -0,0 +1,26 @@
+{
+    public function __construct(
+        protected OvertimeRequestStateManager $stateManager,
+        protected VacationTypeConfigRetriever $configRetriever,
```
Is it used anywhere?

kamilpiech97 replied:
removed

krzysztofrewak commented on app/Domain/OvertimeCalculator.php:
```diff
@@ -0,0 +1,18 @@
+{
+    public function calculateHours(CarbonInterface $from, CarbonInterface $to): int
+    {
+        $hours = Carbon::create($from)->diffInMinutes($to, true);
```
Please use named argument in `true`.

krzysztofrewak commented on app/Enums/SettlementType.php:
```diff
@@ -0,0 +1,28 @@
+
+    public static function casesToSelect(): array
+    {
+        $cases = collect(SettlementType::cases());
+
+        return $cases->map(
+            fn(SettlementType $enum): array => [
+                "label" => $enum->label(),
+                "value" => $enum->value,
+            ],
+        )->toArray();
```
Isn't that overkill?

```suggestion
        return array_map(
            fn(SettlementType $enum): array => [
                "label" => $enum->label(),
                "value" => $enum->value,
            ],
            SettlementType::cases(),
        )
```

krzysztofrewak commented on app/Http/Controllers/OvertimeRequestController.php:
```diff
@@ -0,0 +1,214 @@
+            ->states(OvertimeRequestStatesRetriever::filterByStatusGroup($status, $request->user()))
+            ->paginate();
+
+        $pending = $request->user()
+            ->overtimeRequests()
+            ->whereBelongsTo($yearPeriodRetriever->selected())
+            ->states(OvertimeRequestStatesRetriever::pendingStates())
+            ->count();
+
+        $success = $request->user()
+            ->overtimeRequests()
+            ->whereBelongsTo($yearPeriodRetriever->selected())
+            ->states(OvertimeRequestStatesRetriever::successStates())
+            ->count();
+
+        $failed = $request->user()
+            ->overtimeRequests()
+            ->whereBelongsTo($yearPeriodRetriever->selected())
+            ->states(OvertimeRequestStatesRetriever::failedStates())
+            ->count();
+
+        $settled = $request->user()
+            ->overtimeRequests()
+            ->whereBelongsTo($yearPeriodRetriever->selected())
+            ->states(OvertimeRequestStatesRetriever::settledStates())
+            ->count();
```
Maybe we could cache it somehow?

kamilpiech97 replied:
added query cache, for vacation requests too

krzysztofrewak commented on app/Http/Controllers/OvertimeRequestController.php:
```diff
@@ -0,0 +1,214 @@
+    public function indexForApprovers(
+        Request $request,
+        YearPeriodRetriever $yearPeriodRetriever,
+    ): RedirectResponse|Response {
```
Maybe throwing an exception with redirect would be better?

kamilpiech97 replied:
added `abort(403)`

krzysztofrewak commented on app/Http/Resources/OvertimeRequestResource.php:
```diff
@@ -0,0 +1,52 @@
+            "name" => $this->name,
+            "user" => new SimpleUserResource($this->user),
+            "state" => $this->state,
+            "from" => $this->from->format("Y-m-d H:i"),
```
Maybe you could move this very specific format (`""Y-m-d H:i"`) to some helper/constant?

kamilpiech97 replied:
added Helper

EwelinaSkrzypacz reviewed (commented):
1. When I create an overtime request, the number of hours is always 0 (when I use localhost address, because when I'm using `http://toby.blumilk.localhost/overtime/requests/create` it is fine

![image](https://github.com/blumilksoftware/toby/assets/56546832/59b8a199-c307-408c-9040-c22ccbb39bd2)

2. Please fix this validation massage, because it's a bit unclear.

![image](https://github.com/blumilksoftware/toby/assets/56546832/0e8b07b9-73be-44d5-935e-1df1ec7e2519)

3. For me default time it's shouldn't be midnight - I think that most of overtime is afternoon, so maybe it should be like 16:00? Or 15:00? 

4. As an administrator, I don't think we need "Moje nadgodziny" - maybe just do like we have requests? I mean we have just "Requests". 

![image](https://github.com/blumilksoftware/toby/assets/56546832/c67394fd-3dd8-4bfc-8440-cd1cd46668fd)

5. Date format - in all appliction we have dd.mm.yyy

![image](https://github.com/blumilksoftware/toby/assets/56546832/d5eead3f-f385-4d09-b3d4-b8678083b7eb)

6. As an employee should I be able to add overtime request, when I have pending/accepted vacation?

7. I think we should add here counter, when user has some overtime to accept

![image](https://github.com/blumilksoftware/toby/assets/56546832/ab7bb768-7749-41a3-8734-07acf7fa3303)

8. "Rozliczenie nadgodzin oznaczy nadgodziny jako rozliczone." - can we write this somehow different? 

9. I think that admin should have possibility to cancel settled overtime; just in case - like we have canceling approved vacation requests.

10. Now numeration for overtime and requests looks the same. Maybe we should distinguish it somehow?

11. As I understand accepting/declining overtime in Slack will be implemented in another pr? If so, please create issue for that or at least add comment in issue - I don't want to forget about it.

12. For now, as an administrative approver I can accept overtime. I thought that only technical approver/admin can do that.

kamilpiech97 commented:
> 1. When I create an overtime request, the number of hours is always 0 (when I use localhost address, because when I'm using `http://toby.blumilk.localhost/overtime/requests/create` it is fine
> 
> ![image](https://private-user-images.githubusercontent.com/56546832/340600494-59b8a199-c307-408c-9040-c22ccbb39bd2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3NzQ3ODIsIm5iZiI6MTcxODc3NDQ4MiwicGF0aCI6Ii81NjU0NjgzMi8zNDA2MDA0OTQtNTliOGExOTktYzMwNy00MDhjLTkwNDAtYzIyY2NiYjM5YmQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDA1MjEyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk4MzYwNDg0NmUwYmMzZWNmYzZiYTVmZWIzMDI0NTkyMTQxNzU4YWUzZDkxYjE5NTk5MzRkODNiOTM5NTRhMjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ajuISLD2-jTzZKe4bWvApWXiRnn8Kz7mR8_XR109p7k)
> 
> 2. Please fix this validation massage, because it's a bit unclear.
> 
> ![image](https://private-user-images.githubusercontent.com/56546832/340602011-0e8b07b9-73be-44d5-935e-1df1ec7e2519.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3NzQ3ODIsIm5iZiI6MTcxODc3NDQ4MiwicGF0aCI6Ii81NjU0NjgzMi8zNDA2MDIwMTEtMGU4YjA3YjktNzNiZS00NGQ1LTkzNWUtMWRmMWVjN2UyNTE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDA1MjEyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwNTAwNzViNGMyMGExYWUwNzI4NmVhZDE2YTlhMDI0ZjUzY2Q4MTYxZjM4N2EzZGQ5OGVmMjY4M2M2OWIxYTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KnDAdNtCtlL0Wt6BjTr1yh3Nz41rMgu9b5Y84rHzEtA)
> 
> 3. For me default time it's shouldn't be midnight - I think that most of overtime is afternoon, so maybe it should be like 16:00? Or 15:00?
> 4. As an administrator, I don't think we need "Moje nadgodziny" - maybe just do like we have requests? I mean we have just "Requests".
> 
> ![image](https://private-user-images.githubusercontent.com/56546832/340603498-c67394fd-3dd8-4bfc-8440-cd1cd46668fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3NzQ3ODIsIm5iZiI6MTcxODc3NDQ4MiwicGF0aCI6Ii81NjU0NjgzMi8zNDA2MDM0OTgtYzY3Mzk0ZmQtM2RkOC00YmZjLTg0NDAtY2QxY2Q0NjY2OGZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDA1MjEyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmMThkNjY2NmQzZTRhYmYzMjM1OTAwYzIzM2IzOTQwZGJmMzE0MGE1MDViMGRmNzhlYmFiY2EyMDlhZGMyODUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.s5dtTs5cxVRudMZMSm5C0B3ACSKNU0UC15SRdgRDeBI)
> 
> 5. Date format - in all appliction we have dd.mm.yyy
> 
> ![image](https://private-user-images.githubusercontent.com/56546832/340604539-d5eead3f-f385-4d09-b3d4-b8678083b7eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3NzQ3ODIsIm5iZiI6MTcxODc3NDQ4MiwicGF0aCI6Ii81NjU0NjgzMi8zNDA2MDQ1MzktZDVlZWFkM2YtZjM4NS00ZDA5LWIzZDQtYjg2NzgwODNiN2ViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDA1MjEyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwNDZjMTgyZmU4ZGMwMzQyNDNiNzkzMWVjMzllNGQyMDNkYzg2ZmUwOThhM2EzNzFiMzEyMzJiMjY2MzAyYWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YuSSSQhEiI8zXlHsjjXIFqPlvfz9GTf69Sl7lc2v9cY)
> 
> 6. As an employee should I be able to add overtime request, when I have pending/accepted vacation?
> 7. I think we should add here counter, when user has some overtime to accept
> 
> ![image](https://private-user-images.githubusercontent.com/56546832/340606802-ab7bb768-7749-41a3-8734-07acf7fa3303.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3NzQ3ODIsIm5iZiI6MTcxODc3NDQ4MiwicGF0aCI6Ii81NjU0NjgzMi8zNDA2MDY4MDItYWI3YmI3NjgtNzc0OS00MWEzLTg3MzQtMDdhY2Y3ZmEzMzAzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDA1MjEyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1MTQ2MmRhMDViNGQ0YmY4Yzc5ZjFkNTY5MDVmMDFlMzM3MDAxYmExODkzYmMxNzg0NGUyYjkwYjk4ODFjYjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jqKd22_M-pYmpRsxrCD0_4T1wBVZ97zc6wi0-o2zzzQ)
> 
> 8. "Rozliczenie nadgodzin oznaczy nadgodziny jako rozliczone." - can we write this somehow different?
> 9. I think that admin should have possibility to cancel settled overtime; just in case - like we have canceling approved vacation requests.
> 10. Now numeration for overtime and requests looks the same. Maybe we should distinguish it somehow?
> 11. As I understand accepting/declining overtime in Slack will be implemented in another pr? If so, please create issue for that or at least add comment in issue - I don't want to forget about it.
> 12. For now, as an administrative approver I can accept overtime. I thought that only technical approver/admin can do that.

1. Lets create issue for that.
2. fixed
3. I will stay with that
4. Fixed
5. Fixed
6. Let's ask Kamila but AFAIK we do not want create compliated valdation for overtime process
7. Added
8. Fixed
9. Added
10. Changed
11. In next PR
12. I will stay with that

EwelinaSkrzypacz reviewed (commented):
When I logged as an employee and click "Moje nadgodziny" I see error:

![image](https://github.com/blumilksoftware/toby/assets/56546832/f1310726-8cd7-4d0c-b3b7-ff85ab0d600e)

kamilpiech97 commented:
> When I logged as an employee and click "Moje nadgodziny" I see error:
> 
> ![image](https://private-user-images.githubusercontent.com/56546832/341016801-f1310726-8cd7-4d0c-b3b7-ff85ab0d600e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3OTMyMTgsIm5iZiI6MTcxODc5MjkxOCwicGF0aCI6Ii81NjU0NjgzMi8zNDEwMTY4MDEtZjEzMTA3MjYtOGNkNy00ZDBjLWIzYjctZmY4NWFiMGQ2MDBlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE5VDEwMjgzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU4YjY1OTMxMjFiMWQ2OWQzYzVlNzY5YjE3Zjg2NTdiMjFjNTA1YmYyZjA5MWVhMmRmOGVlMjE1NTkxYjA2Y2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZRONfX3Ym9mQlG157yPNvLZeet97xJRobAqHB7Mer_c)

Run seeder?

EwelinaSkrzypacz reviewed (commented):
1. As a employee I can settle overtime. Is it okay for us?
2. When I click "Moje nadgodziny" in menu and click button "Dodaj wniosek" (so I'm on the page with creating overtime) on the menu the text "Moje nadgodziny" is not highlighted - in requests it is.
3. So, when I have settled overtime for example 19.06.2024 15:00 - 19.06.2024 17:00, I can create another overtime 19.06.2024 16:00 - 19.06.2024 18:00. Is this okay for us?

kamilpiech97 commented:
> 1. As a employee I can settle overtime. Is it okay for us?
> 2. When I click "Moje nadgodziny" in menu and click button "Dodaj wniosek" (so I'm on the page with creating overtime) on the menu the text "Moje nadgodziny" is not highlighted - in requests it is.
> 3. So, when I have settled overtime for example 19.06.2024 15:00 - 19.06.2024 17:00, I can create another overtime 19.06.2024 16:00 - 19.06.2024 18:00. Is this okay for us?

1. yep
2. fixed
3. fixed, added edge cases

krzysztofrewak commented on app/Http/Controllers/OvertimeRequestController.php:
```diff
@@ -0,0 +1,223 @@
+            ->overtimeRequests()
+            ->whereBelongsTo($yearPeriodRetriever->selected())
+            ->states(OvertimeRequestStatesRetriever::failedStates())
+            ->cache(key: "overtimeStats")
```
Does it cache it globally or per user?

kamilpiech97 replied:
globally but only for 60 seconds

kamilpiech97 replied:
it is for index page where u switching between tabs with specific states
